_posts/2017/2017-04-19-build-a-multiple-repository-pull-request-test.md

```diff
@@ -0,0 +1,161 @@
+
+How can we now ensure, that the PR is only triggered once and not for every repository, where a Pull request was set from a developer?
+
+The solution is to build a hierarchy into the repository structure.
+As every repository has a describing [Jenkinsfile](https://jenkins.io/doc/book/pipeline/jenkinsfile/), we need a kind of control repository.
+
+{% image blog/blog-pr-test-1.png %}
+
+This is important to avoid uncontrollable starts of the PR_Test.
+If a Pull request exists there, the other repositories don't have to activate their tests as the control repository will do it.
+
+{% image blog/blog-pr-test-2.png %}
+
+Other repositories have to check, if a branch with the given name exist in the control repository.
+If the branch was found, the control repository will start the test.
+If the branch don't exist, one of the other repositories has to trigger the PR test.
+
+{% image blog/blog-pr-test-3.png %}
+
+One way or another, we will start to build a new Job called 'PR_Test'.
+
+````
+    ...
+    // Evaluate, if branches exist in the given repository.
+    if (git.getBranchForRepository("multistore-MS", BRANCH, REMOTE)) {
+        // If a branch with the same name exist in control repository, we can exit here,
+        // as the PR test will be triggered by Jenkinsfile of control repository
+        println "We exit here. Control repository will take care of the PR test..."
+        currentBuild.result = 'SUCCESS'
+        return
+    } else {
+        // Set the branch to the standard, if branch do not exist.
+        println "We can use dev for multistore-MS..."
+        EMS_BRANCH_NAME = "dev"
+    }
+    ...
+````
+```
+    ...
+    // Trigger the PR test
+    build job: "PR_Test", wait: false, parameters: [
+          [$class: 'StringParameterValue', name: 'ep6Branch', value: EP6_BRANCH_NAME],
+          [$class: 'StringParameterValue', name: 'ep6Remote', value: EP6_REMOTE_NAME],
+          [$class: 'StringParameterValue', name: 'emsBranch', value: EMS_BRANCH_NAME],
+          [$class: 'StringParameterValue', name: 'emsRemote', value: EMS_REMOTE_NAME],
+          [$class: 'StringParameterValue', name: 'rtgBranch', value: RTG_BRANCH_NAME],
+          [$class: 'StringParameterValue', name: 'rtgRemote', value: RTG_REMOTE_NAME]]
+    ...
+```
+
+### 3. Environment set up, how to start the test?
+
+To be able to work with different repositories, you need a global job as pull request job.
+It is easy to start from different repositories, as you only have to overhand variables like branch and remote name.
+Beginning with the tests, we have to set up our environment first.
+This will be done in parallel.
+It can be seen as the first test to assure that at least the installation is still functional.
+After that, your tests should run as expected.
+Beginning with short running tests like unit tests up to long running tests like UI tests (if there is a UI).
```

Comment:
Beginning with short unit tests up to longer tests such as UI tests (if a UI exists).